inc/classes/subscriber/Optimization/class-lazyload-subscriber.php

```diff
@@ -426,6 +427,22 @@ public function deactivate_lazyload_on_specific_posts() {
 		}
 	}
 
+	/**
+	 * Disable WP core lazyload if our images lazyload is active
+	 *
+	 * @since 3.5
+	 * @author Remy Perona
+	 *
+	 * @return bool
+	 */
+	public function maybe_disable_core_lazyload() {
+		if ( $this->can_lazyload_images() ) {
+			return false;
+		}
+
+		return true;
```

Comment:
This code can be simplified to return the inverse value of the conditional expression:

```php
return ! (bool) $this->can_lazyload_images();
```

Reply:
@hellofromtonya Is there a purpose to this `(bool)`?

Reply:
The method `can_lazyload_images()` returns a value through a filter. Therefore, it can return mixed values if some code is hooked into it and returns say a `1` or '0'` or something else. the (bool) here ensures we have a boolean value that is being inverted.

Reply:
Isn't the `!` already insuring it? The boolean cast seems redundant to me.

Reply:
With PHP 7, we want to be strict with our data types. Here we are expecting a boolean and then we invert it. But what if we get a string or an integer. Yes, PHP can type juggle, but with PHP 7 we can strive to be very strict with what data type we expect.

Does that make sense?

Reply:
Still not. Is there a case where this doesn't return a boolean?
```php
return ! $this->can_lazyload_images();
```
For me, `!` will always cast the result as a boolean, am I wrong?

Reply:
Yes. It's running through a filter. So any callback can return any value and data type.

```php
	/**
	 * Checks if we can lazyload images.
	 *
	 * @since 3.3
	 *
	 * @return boolean
	 */
	private function can_lazyload_images() {
		if ( ! $this->options->get( 'lazyload' ) ) {
			return false;
		}

		/**
		 * Filters the lazyload application on images
		 *
		 * @since 2.0
		 * @author Remy Perona
		 *
		 * @param bool $do_rocket_lazyload True to apply lazyload, false otherwise.
		 */
		return apply_filters( 'do_rocket_lazyload', true ); // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
	}
```

The way to ensure we always get a boolean is to put the `(bool)` in that method like this:

```php
return (bool) apply_filters( 'do_rocket_lazyload', true ); // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals
```

Reply:
I think we don't understand each other.
I agree that a value which is filtered can return any type of data, and that we must cast it as boolean to insure we get the right format, I do it all the time. What I mean is that `!` is already doing this job, so there is no need to add another `(bool)` cast behind it.
This:
```php
return ! (bool) $this->can_lazyload_images();
```
Does exactly the same thing than this:
```php
return ! $this->can_lazyload_images();
```
Here, `(bool)` is useless.

Reply:
>What I mean is that ! is already doing this job, so there is no need to add another (bool) cast behind it.

When the value come back from the method is not a boolean, then yes, `!` is type juggling that value into a boolean and then inverting it.

For example, let's say `$this->can_lazyload_images();` returns a string data type with a value of `'some value'`. What's happening here?

1. With `(bool)` we are type casting to ensure it's a boolean. We are being strict with our intent of the data type.
2. Then the `!` inverts the boolean.

What happens if we don't have the type casting to boolean data type? In PHP 5 (since PHP Is a weakly typed language) we are used to letting these operators do the type juggling to convert it for us into whatever the data type is supposed to be for the operation to continue.

With PHP 7 though, we are moving more and more away from weakly typed where possible and striving to be more purposeful about our data types.

So to answer your question:

1. Yes, if we stick with weakly typed, then yes, PHP will convert whatever is returned from the method into a boolean for us, i.e. without having to use explicitly tell PHP we want a boolean.
2. However, if we want to be more explicit and strict, then we tell PHP (and ourselves) what the data type should be.

Reply:
A better solution here is to ensure a method always returns a known data type. In this case, `can_lazyload_images()` should return a boolean. Then there's no need for explicitly doing type casting like I showed above.

Reply:
OK, now I understand the purpose of `(bool)`: is not to prevent type juggling on the method’s output, but on `!`.
Wow, I don't think I like this, but it's another matter.
Thanks for the explanation.

Reply:
You're welcome. Thanks for discussing it with me. Always enjoy a good technical discussion!